File: Leetcode/stack/503_NextGreaterElementII/Solution.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        if(nums==null || nums.length==0) return nums;
        int n = nums.length;
        int[] res = new int[n];
        Arrays.fill(res, -1);
        Stack<Integer> stack = new Stack<>();
        for(int i = 0; i<2*n-1; ++i){
            while(!stack.empty() && nums[stack.peek()%n]<nums[i%n])
                res[stack.pop()] = nums[i%n];
            stack.push(i%n);
        }
        return res;
    }
}

File: Leetcode/stack/503_NextGreaterElementII/Solution.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        if n==0: return nums
        res = [-1]*n
        stack = []
        for i in range(2*n-1):
            while stack and nums[stack[-1]%n] < nums[i%n]:
                key = stack.pop()
                res[key] = nums[i%n]
            stack.append(i%n)
        return res
